### PR TITLE
[in_app_purchase_android] Implement `BillingClient` connection management and introduce `BillingClientManager`

### LINKED ISSUE #110663: [in_app_purchase][android] `BillingClient` connection is not restored after it's lost
As described in [Google Play Billing docs](https://developer.android.com/google/play/billing/integrate#connect), connection to the `BillingClient` may be lost at any time and apps making use of this API should implement their own connection management logic.
App is notified about such a disconnection via [`BillingClientStateListener.onBillingServiceDisconnected()`](https://developer.android.com/reference/com/android/billingclient/api/BillingClientStateListener?hl=en#onBillingServiceDisconnected()) and/or [`BillingClientResponseCode.SERVICE_DISCONNECTED`](https://developer.android.com/reference/com/android/billingclient/api/BillingClient.BillingResponseCode?hl=en#SERVICE_DISCONNECTED)

Currently, `InAppPurchaseAndroidPlatform` ignores `onBillingServiceDisconnected` callback and does not handle `BillingResponse.serviceDisconnected`:

https://github.com/flutter/plugins/blob/35372527c70ed0ae4cc5392330f53ede9774d88a/packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform.dart#L222-L223

https://github.com/flutter/plugins/blob/35372527c70ed0ae4cc5392330f53ede9774d88a/packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform.dart#L154-L176

To be consistent with iOS, Android implementation of `InAppPurchase` should manage the `BillingClient` connection transparently, so that a operation never fails because of `BillingClient` disconnection.

In my own experience, frequency of such disconnects varies between devices. On one of the devices I've tested, connection was lost every time the app was paused. Since starting the purchase flow opens system payment UI (which pauses the app), when user made the payment and came back to the app, `completePurchase` method failed silently because of the lost connection. This, in turn, caused every purchase to be eventually refunded due to lack of acknowledgement.

This issue was also reported in https://github.com/flutter/flutter/issues/40054

## Steps to Reproduce

1. Run any app that makes use of `in_app_purchase` on Android (e.g. [official `in_app_purchase` example](https://github.com/flutter/plugins/tree/main/packages/in_app_purchase/in_app_purchase/example))
2. Perform any action that makes use of `BillingClient` API (e.g. start - and then possibly cancel - a purchase flow)
3. Simulate disconnection of `BillingClient` by executing `adb shell pm clear com.android.vending` on the host machine (`Billing service disconnected.` should appear in logcat)
4. Perform any action that makes use of `BillingClient` API (as in step 2.)

Expected results: Action completes with success

Actual results: Action fails (e.g. in case of `queryProductDetails` call, empty list is returned)

Reproduced on:
- Flutter 3.0.4
- `in_app_purchase` 3.0.7
- `in_app_purchase_android` 0.2.3+3

`flutter doctor -v`:
```
[✓] Flutter (Channel stable, 3.0.4, on Ubuntu 22.04.1 LTS 5.15.0-46-generic, locale pl_PL.UTF-8)
    • Flutter version 3.0.4 at /home/kuba/snap/flutter/common/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 85684f9300 (9 weeks ago), 2022-06-30 13:22:47 -0700
    • Engine revision 6ba2af10bb
    • Dart version 2.17.5
    • DevTools version 2.12.2

[!] Android toolchain - develop for Android devices (Android SDK version 32.0.0-rc1)
    • Android SDK at /home/kuba/tools/android-sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/linux#android-setup for more details.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
    • cmake version 3.10.2
    • ninja version 1.8.2
    • pkg-config version 0.29.1

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[✓] IntelliJ IDEA Ultimate Edition (version 2022.2)
    • IntelliJ at /home/kuba/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/222.3739.54
    • Flutter plugin version 69.0.5
    • Dart plugin version 222.3739.24

[✓] Connected device (3 available)
    • ANE LX1 (mobile) • 9WV7N18426037007 • android-arm64  • Android 9 (API 28)
    • Linux (desktop)  • linux            • linux-x64      • Ubuntu 22.04.1 LTS 5.15.0-46-generic
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 102.0.5005.61

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 2 categories.
```

### Logs

Logs from reproduction using my own app using `in_app_purchase` on Android:

```
I/flutter (18225): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter (18225): │ 💡 [SubscriptionService] Product details found, starting purchase: 
I/flutter (18225): │ 💡 title=Standard (haveagood.one)
I/flutter (18225): │ 💡 price=9,99 zł
I/flutter (18225): │ 💡 description=
I/flutter (18225): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
...omitted...
W/BillingClient(18225): Billing service disconnected.
...omitted
I/flutter (18225): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter (18225): │ Exception: Product details not found for standard1
I/flutter (18225): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
...omitted...
I/flutter (18225): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter (18225): │ ⛔ Uncaught error
I/flutter (18225): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

### PR DESCRIPTION
**Pull request by SynSzakala**
This PR adds management of `BillingClient` connection to Android implementation of `InAppPurchase`. This should fix issues caused by lost connections.

New and previously existing connection management is extracted to a new component, [`BillingClientManager`](https://github.com/flutter/plugins/pull/6309/files#diff-29a412a27432e3733962c0cdbe45bc130501148f6d3d4a4fd4e657c4e5832bb8R24). 
This component is exported in [`billing_client_wrappers.dart`](https://github.com/flutter/plugins/pull/6309/files#diff-25155199c901c01755b6452184f2bc0af729c3fc2439eeb582983d17ca14a019R5), so that it can also be used when interacting with `BillingClient` directly.

Fixes https://github.com/flutter/flutter/issues/110663

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by dgjamaster** [comment]
I look forward to this PR being merged.
Thanks!!

**Comment by stuartmorgan** [comment]
@GaryQian Ping on this review.

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart:1]: You should document in BillingClient the drawbacks of using the BillingClient class directly and point them to the newly recommended way of handling operations using this new BillingClientManager class.

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart:97]: Document when/under what conditions dispose should be called.

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/sku_details_wrapper.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/sku_details_wrapper.dart:23]: Is this class in the correct file? I expected this to live in the billing_client_manager.dart file.

File: packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform.dart:160]: Avoid unnecessary refactors

File: packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform.dart:236]: Avoid unnecessary refactors

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart:41]: This sentence is redundant as @visiblefortesting will enforce preventing its usage outside of tests.

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart:62]: nit: Add a newline

File: packages/in_app_purchase/in_app_purchase_android/test/billing_client_wrappers/billing_client_manager_test.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/test/billing_client_wrappers/billing_client_manager_test.dart:52]: nit: newline between tests

File: packages/in_app_purchase/in_app_purchase_android/test/billing_client_wrappers/billing_client_manager_test.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/test/billing_client_wrappers/billing_client_manager_test.dart:66]: nit: newline between tests

File: packages/in_app_purchase/in_app_purchase_android/test/billing_client_wrappers/billing_client_manager_test.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/test/billing_client_wrappers/billing_client_manager_test.dart:74]: nit: newline between tests

File: packages/in_app_purchase/in_app_purchase_android/test/billing_client_wrappers/billing_client_manager_test.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/test/billing_client_wrappers/billing_client_manager_test.dart:94]: nit: newline between tests

File: packages/in_app_purchase/in_app_purchase_android/test/in_app_purchase_android_platform_test.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/test/in_app_purchase_android_platform_test.dart:69]: nit: newline between tests

File: packages/in_app_purchase/in_app_purchase_android/test/in_app_purchase_android_platform_test.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/test/in_app_purchase_android_platform_test.dart:62]: nit: newline between tests

File: packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/sku_details_wrapper.dart
Comment (GaryQian) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/sku_details_wrapper.dart:8]: Prefer directly importing billing_client_manager.dart over the export file which contains extraneous imports.

**Comment by GaryQian** [comment]
@SynSzakala Overall looks good, just a few small nits. Let me know when you address them!

**Comment by GaryQian** [comment]
@SynSzakala It looks like there are some CLA signing issues where your secondary email needs to accept the CLA, or you can amend the commit that used your alternate email address and force-push.

**Review by GaryQian — APPROVED** [review]
LGTM! Thanks for your contribution. Once you address the CLA issue, we can move forwards.

**Comment by SynSzakala** [comment]
@GaryQian Everything should be ready for merge now.

**Comment by olof-dev** [comment]
Thanks for this PR! The issue with lost connections leading to refunded purchases is hitting us pretty hard in production, so this is much appreciated.

**Comment by SynSzakala** [comment]
@stuartmorgan @GaryQian Is anything preventing us from merging this?

File: packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform.dart
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/lib/src/in_app_purchase_android_platform.dart:256]: Why have you added commas everywhere? Changing formatting of lines that you aren't otherwise changing makes the diff (and git blame) more confusing.

**Review by stuartmorgan-g — COMMENTED** [review]
@SynSzakala Could you address all of the open comments from @GaryQian's review, and fix the analysis issues that are causing `repo_checks` to fail?

Once that's done I can approve this for landing.

**Comment by GaryQian** [comment]
@SynSzakala When you get a chance to fix the failing analyzer checks, we can move forwards with this. Thanks!

**Comment by olof-dev** [comment]
@SynSzakala It looks like the analyzer complaints are just warnings to do with 4 unneeded `await`s. I was gonna send you a PR that deletes them, but I realized that there might then be CLA-type stuff to deal with. In any case, thanks again for this PR - it's substantial!

**Comment by SynSzakala** [comment]
@GaryQian @stuartmorgan Cleared all comments (including unnecessary refactors), fixed analyzer issues. Looking forward to merge this!

**Review by GaryQian — APPROVED** [review]
LGTM! Thanks

**Comment by SynSzakala** [comment]
@stuartmorgan @GaryQian Are there any additional steps needed to merge this? I don't have the access to add "autosubmit" label as suggested in [Tree Hygiene](https://github.com/flutter/flutter/wiki/Tree-hygiene)

**Comment by stuartmorgan** [comment]
> Are there any additional steps needed to merge this?

This needs a second review, per https://github.com/flutter/flutter/wiki/Tree-hygiene#getting-a-code-review

@bparrishMines Could you do the second review here, as the ecosystem-side codeowner for this plugin?

**Comment by stuartmorgan-g** [comment]
We've just completed the migration of the plugin code to the flutter/packages repository, as described in https://flutter.dev/go/flutter-plugins-repo-migration, and this repository is now being archived. Unfortunately that means that all in-progress PRs here must be moved to flutter/packages.

Please see [our instructions](https://github.com/flutter/flutter/wiki/Migrating-Plugins-repository-PRs-to-Packages) for an explanation of how to move your PR, and if you have any issues moving your PR please don't hesitate to reach out in the #hackers-ecosystem channel in [Discord](https://github.com/flutter/flutter/wiki/Chat).

Our apologies that your PR was caught in this one-time transition. We're aware that it's disruptive in the short term, and appreciate your help in getting us to a better long-term state!

**Pull request closed without merging**
